src/allocator.rs

```diff
@@ -0,0 +1,232 @@
+        let ptr = self.0.as_ptr();
+
+        let layout = unsafe { (*ptr).layout };
+        unsafe { alloc::dealloc(ptr as *mut _, layout) };
+    }
+
+    pub fn as_mut(&mut self) -> &mut [u8] {
+        self.as_mut_and_next().0
+    }
+
+    pub fn mut_next(&mut self) -> &mut Option<Buffer> {
+        self.as_mut_and_next().1
+    }
+
+    pub fn as_mut_and_next(&mut self) -> (&mut [u8], &mut Option<Buffer>) {
+        // Safety: TODO()
+        let as_mut = unsafe { &mut (*self.0.as_ptr()).payload };
+        let next = unsafe { &mut (*self.0.as_ptr()).next };
+
+        (as_mut, next)
+    }
+
+    fn repr_c_layout(fields: &[Layout]) -> Result<Layout, LayoutError> {
+        let mut layout = Layout::from_size_align(0, 1)?;
+        for &field in fields {
+            let (new_layout, _) = layout.extend(field)?;
+            layout = new_layout;
+        }
+        Ok(layout.pad_to_align())
+    }
+
+    fn compute_layout(layout_size: usize) -> Result<Layout, LayoutError> {
+        let header_layout = Self::repr_c_layout(Self::HEADER_BUFFER_FIELDS)?;
+        let (layout, _) =
+            header_layout.extend(Layout::for_value(vec![0u8; layout_size].as_slice()))?;
+
+        Ok(layout.pad_to_align())
+    }
+
+    fn layout(&self) -> Layout {
+        unsafe { (*self.0.as_ptr()).layout }
+    }
+
+    fn as_ptr(&self) -> *const BufferLayout {
+        self.0.as_ptr()
+    }
+}
+
+type Sender = mpsc::Sender<Buffer>;
+type Reciever = mpsc::Receiver<Buffer>;
+
+struct BufferChain {
+    buffer: Option<Buffer>,
+    sender: Sender,
+}
+
+impl BufferChain {
+    pub fn new(sender: Sender) -> Self {
+        Self { buffer: None, sender }
+    }
+
+    pub fn push(&mut self, mut buffer: Buffer) {
+        assert!((*buffer.mut_next()).is_none());
+
+        *buffer.mut_next() = self.buffer.take();
+        self.buffer = Some(buffer)
+    }
+
+    // TODO(i.erin) implement Iterator for that
+    pub fn to_vec(Self { buffer, .. }: &mut Self) -> Vec<&mut [u8]> {
+        let mut result = Vec::new();
+
+        let mut current = buffer;
+        loop {
+            match current {
+                Some(buffer) => {
+                    let (as_mut, next) = buffer.as_mut_and_next();
+                    current = next;
+
+                    result.push(as_mut)
+                }
+                None => return result,
+            }
+        }
+    }
+
+    pub fn blocking_dealloc(Self { buffer: mut next_buffer, sender }: Self) {
+        while let Some(mut buffer) = next_buffer {
+            next_buffer = buffer.mut_next().take();
+            sender.blocking_send(buffer).expect("can't send buffer");
+        }
+    }
+
+    pub async fn dealloc(Self { buffer: mut next_buffer, sender }: Self) {
+        while let Some(mut buffer) = next_buffer {
+            next_buffer = buffer.mut_next().take();
+            sender.send(buffer).await.expect("can't send blocking");
+        }
+    }
+}
+
+struct Allocator {
+    reciever: mpsc::Receiver<Buffer>,
```

Comment:
Types are already defined
```suggestion
    Receiver,
```